### PR TITLE
fix(redis): fix dlq logic in unacked retry handler

### PR DESCRIPTION
**Pull request by asher**

### CODE REVIEW

File: keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt [hunk 1]
```diff
@@ -295,27 +295,23 @@ class RedisQueue(
       zrem(unackedKey, fingerprint)
       hdel(messagesKey, fingerprint)
       del("$locksKey:$fingerprint")
-
-      // TODO: use AttemptAttribute instead
-      hdel(attemptsKey, fingerprint)
     }
   }
 
   private fun Jedis.readMessageWithoutLock(fingerprint: String, block: (Message) -> Unit) {
```
Comment (asher) [inline keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt]: `readMessageWithoutLock()` is now a read-only method.

File: keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt [hunk 2]
```diff
@@ -176,35 +177,34 @@ class RedisQueue(
 
           fingerprints.forEach { fingerprint ->
             val attempts = redis.hgetInt(attemptsKey, fingerprint)
-            if (attempts >= Queue.maxRetries) {
```
Comment (asher) [inline keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt]: This resulted in most long running tasks getting deadlettered whenever they ended up unacked, as `attemptsKey` was incremented on every message read. The `attempts` message attribute continues to be incremented as before, but `attemptsKey` is now only incremented by the unacked message retrier, or when an `IOException` is thrown in `readMessage()`.

File: keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt [hunk 3]
```diff
@@ -295,27 +295,24 @@ class RedisQueue(
       zrem(unackedKey, fingerprint)
       hdel(messagesKey, fingerprint)
       del("$locksKey:$fingerprint")
-
-      // TODO: use AttemptAttribute instead
       hdel(attemptsKey, fingerprint)
     }
   }
 
   private fun Jedis.readMessageWithoutLock(fingerprint: String, block: (Message) -> Unit) {
```
[the remark this answers is not in the document]
Comment (robzienert) [reply keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt]: I thought our plan was to get rid of the attempts key in Redis entirely and rely on the attempts attribute on the message?

[the remark this answers is not in the document]
Comment (asher) [reply keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt]: The retry message attribute currently has meaning in the context of orca's RetryableTask, where it tracks number of times a message is returned to the queue for future work. I think there's value in having a separate keiko level unacked retry metric that isn't exposed to userland message handlers.

**Pull request merged**